### ISSUE TITLE
Seemingly incorrect error for missing trait bounds on associated type

### ISSUE THREAD

**Issue by djc** [rust-lang/rust#81162]
I tried this code:

```rust
fn test() {
    let data = read::<[f32; 128]>("foo", 128);
}

fn read<T>(fname: &str, dimension: usize) -> Vec<T>
where
    for<'a> T: TryFrom<&'a [f32]>,
    for<'a> <T as TryFrom<&'a [f32]>>::Error: std::fmt::Debug,
```

I expected to see this happen: this should work, since `[f32; N]` implements `TryFrom<&[f32]>` with `Error = TrySliceError` and `TrySliceError` implements `Debug`.

Instead, this happened:

```rust
error[E0277]: `<[f32; 128] as TryFrom<&'a [f32]>>::Error` doesn't implement `Debug`
  --> tests/foo.rs:7:16
   |
7  |     let data = read::<[f32; 128]>("foo", 128);
   |                ^^^^^^^^^^^^^^^^^^^^^^^^ `<[f32; 128] as TryFrom<&'a [f32]>>::Error` cannot be formatted using `{:?}` because it doesn't implement `Debug`
...
10 | fn read<T>(fname: &str, dimension: usize) -> Vec<T>
   |    ---------- required by a bound in this
...
13 |     for<'a> <T as TryFrom<&'a [f32]>>::Error: std::fmt::Debug,
   |                                               --------------- required by this bound in `read`
   |
   = help: the trait `for<'a> Debug` is not implemented for `<[f32; 128] as TryFrom<&'a [f32]>>::Error`
```

### Meta
`rustc --version --verbose`:
```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-apple-darwin
release: 1.49.0
```

**Comment by estebank** [comment]
Note that not using `for<'a>` makes [the code compile](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=e6ca89df5e0c8ac41d8b2767a8977bd7), but doesn't have the same lifetime semantics.
